src/Stratis.Bitcoin/P2P/Peer/NetworkPeerConnection.cs

```diff
@@ -176,40 +176,32 @@ public async Task ConnectAsync(IPEndPoint endPoint, CancellationToken cancellati
 
             try
             {
-                // This variable records any error occurring in the thread pool task's context.
-                Exception error = null;
-
-                await Task.Run(() =>
-                {
-                    try
-                    {
-                        this.tcpClient.ConnectAsync(endPoint.Address, endPoint.Port).Wait(cancellation);
-                    }
-                    catch (Exception e)
-                    {
-                        // Record the error occurring in the thread pool's context.
-                        error = e;
-                    }
-                }).ConfigureAwait(false);
-
-                // Throw the error within this error handling context.
-                if (error != null)
-                    throw error;
+                this.asyncProvider.Signals.Publish(new PeerConnectionAttempt(false, endPoint));
+                await this.tcpClient.ConnectAsync(endPoint.Address, endPoint.Port).WithCancellationAsync(cancellation).ConfigureAwait(false);
+                this.asyncProvider.Signals.Publish(new PeerConnected(false, endPoint));
 
                 this.stream = this.tcpClient.GetStream();
             }
             catch (OperationCanceledException)
             {
-                this.logger.LogDebug("Connecting to '{0}' cancelled.", endPoint);
+                this.asyncProvider.Signals.Publish(new PeerConnectionAttemptFailed(false, endPoint, "Operation Canceled"));
+                this.logger.LogTrace("Connecting to '{0}' cancelled.", endPoint);
                 this.logger.LogTrace("(-)[CANCELLED]");
                 throw;
             }
-            catch (Exception e)
+            catch (SocketException ex)
+            {
+                this.asyncProvider.Signals.Publish(new PeerConnectionAttemptFailed(false, endPoint, $"Socket Exception: {ex.Message}"));
+                this.logger.LogDebug("Error connecting to '{0}', exception message: {1}", endPoint, ex.Message);
+                this.logger.LogTrace("(-)[UNHANDLED_EXCEPTION]");
+                throw;
+            }
+            catch (Exception ex)
             {
-                if (e is AggregateException) e = e.InnerException;
-                this.logger.LogDebug("Error connecting to '{0}', exception message: {1}", endPoint, e.Message);
+                this.asyncProvider.Signals.Publish(new PeerConnectionAttemptFailed(false, endPoint, ex.Message));
```

Comment:
Why not pass the ex in this case of PeerConnectionAttemptFailed?

Reply:
this is raised during connection and from my tests it doesn't give any information better than the message itself (the message could be very long, look for "socket exception" in the output example I've posted).

Consider that these messages are stored in memory when you enable PeerDiagnostic, so I wanted to preserve memory storing only useful data.

If we see that having the full exception object with its stack trace I'm not against passing the exception itself , anyway note that the message contains already some stack information, see the example I'm talking about:

> "[19/04/2019 12:02:37] Disconnected. Reason: Unexpected exception while connecting to socket. Exception: System.Net.Internals.SocketExceptionFactory+ExtendedSocketException (10061): Impossibile stabilire la connessione. Rifiuto persistente del computer di destinazione [::ffff:185.244.214.216]:16178\r\n   at Stratis.Bitcoin.P2P.Peer.NetworkPeerConnection.ConnectAsync(IPEndPoint endPoint, CancellationToken cancellation) in E:\\Sviluppo\\GitHub-Projects\\StratisBitcoinFullNode\\src\\Stratis.Bitcoin\\P2P\\Peer\\NetworkPeerConnection.cs:line 211\r\n   at Stratis.Bitcoin.P2P.Peer.NetworkPeer.ConnectAsync(CancellationToken cancellation) in E:\\Sviluppo\\GitHub-Projects\\StratisBitcoinFullNode\\src\\Stratis.Bitcoin\\P2P\\Peer\\NetworkPeer.cs:line 415\r\n   at Stratis.Bitcoin.P2P.Peer.NetworkPeer.ConnectAsync(CancellationToken cancellation) in E:\\Sviluppo\\GitHub-Projects\\StratisBitcoinFullNode\\src\\Stratis.Bitcoin\\P2P\\Peer\\NetworkPeer.cs:line 450"

Reply:
Well if you enabled diagnostics you want to have as much info as possible memory is not a factor in such cases.